### ISSUE TITLE
System Error 87 when sharing path from linux

### ISSUE THREAD

**Issue by casualsnek** [casualsnek/cassowary#11]
Sytem error 87 is reported on some system when mapping path on linux to Windows drive

**Comment by casualsnek** [comment]
If anyone can reproduce this issue please get in touch !

**Comment by CanasSimon** [comment]
I am indeed experiencing this exact same problem

On Arch Linux

**Comment by casualsnek** [comment]
Which edition of windows 10 are you running on ? Is it standard windows 10 pro or something else ?

**Comment by CanasSimon** [comment]
Since I followed the guide, yes, I should have the standard Windows 10 Pro, but unactivated (dunno if this changes anything)

**Comment by casualsnek** [comment]
Can you open app using ```cassowary -a``` then try to create a share and allow error to show, then open explorer using ```cassowary -c guest-run -- explorer.exe`` navigate to C:\Users\USERNAME\.config\casualrdh  and upload the log file located there !

**Comment by CanasSimon** [comment]
<details>
<summary>'cassowary -a' log</summary>

```
[ 2022-02-05 14:51:47,920 ] | [  DEBUG ] :  [    __init__ ->                 main  ] -->  Starting configuration GUI 
Icon theme "Mint-X" not found.
[ 2022-02-05 14:51:48,059 ] | [  DEBUG ] :  [      helper ->              vm_wake  ] -->  Attempting to resume VM 
[ 2022-02-05 14:51:48,062 ] | [ WARNING ] :  [      helper ->              vm_wake  ] -->  VM state is not set to suspended : State -> '1'  
[ 2022-02-05 14:51:48,063 ] | [  DEBUG ] :  [      helper ->     fix_black_window  ] -->  Opening & closing a test window to trigger login or try to fix black screen bug on first launch 
[ 2022-02-05 14:51:56,851 ] | [  DEBUG ] :  [      helper ->     fix_black_window  ] -->  Created a marker -> One session done 
[ 2022-02-05 14:51:58,854 ] | [  DEBUG ] :  [      helper ->     fix_black_window  ] -->  Created a marker -> One session done 
[ 2022-02-05 14:52:00,856 ] | [  DEBUG ] :  [      helper ->     fix_black_window  ] -->  An app was already opened, the black window should not appear now ! 
Trying to reconnect
[ 2022-02-05 14:52:00,856 ] | [   INFO ] :  [      client ->      init_connection  ] -->  Attempting to connect to server 
[ 2022-02-05 14:52:01,877 ] | [   INFO ] :  [      client ->      init_connection  ] -->  Connected to server at 192.168.122.59:7220 
[ 2022-02-05 14:52:01,877 ] | [  DEBUG ] :  [      client ->      init_connection  ] -->  Starting sender and receiver threads 
[ 2022-02-05 14:52:04,112 ] | [  DEBUG ] :  [      client ->               __send  ] -->  Sending message to server 
[ 2022-02-05 14:52:15,798 ] | [  DEBUG ] :  [      client ->               __send  ] -->  Sending message to server
```
</details>
<details>
<summary>casualrdh.log</summary>

```
[ 2022-02-05 03:06:34,601 ] | [  ERROR ] :  [    cmd_asso ->             __init__  ] -->  Ftype not associated to app, dobule clicking file will not trigger xdg-open 
[ 2022-02-05 03:18:16,187 ] | [  ERROR ] :  [    cmd_dirs ->             __init__  ] -->  Failed to map host root to Z: -> net use command returned: System error 87 has occurred.

The parameter is incorrect.

 
[ 2022-02-05 14:09:09,708 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:09:19,411 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:09:53,030 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:09:56,998 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:11:00,643 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:11:04,409 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:11:07,877 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:11:17,506 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:11:21,531 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:11:33,111 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:13:09,074 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:13:13,308 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:13:15,339 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:13:18,933 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:19:04,333 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:19:22,520 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:21:28,111 ] | [ WARNING ] :  [    cmd_asso ->              run_cmd  ] -->  Error Unsetting file association, Command - ['unset-association', 'dgo']  : Traceback (most recent call last):
  File "base\command\cmd_asso.py", line 139, in run_cmd
  File "base\command\cmd_asso.py", line 91, in __unset_associations
IndexError: list index out of range
 
[ 2022-02-05 14:21:39,455 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:21:41,939 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:22:20,013 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
[ 2022-02-05 14:52:16,603 ] | [  ERROR ] :  [    cmd_dirs ->        __add_new_map  ] -->  Unknown error while trying to map to drive: 'System error 87 has occurred.

The parameter is incorrect.

' 
```
</details>

I don't have a disk called "Z:" but I do have one that's called 'root on [username]' which I guess is the same thing?

**Comment by casualsnek** [comment]
Looks like mapping is failing for some, i will work on a fix asap !

**Comment by sugguy** [comment]
First of all thank you for this great project.

I figured out a way to fix it : 
TL;DR : remove all mapped drivers using powershell command and then remap.
How to do it step by step:
1. add powershell from guest apps in cassowary. and launch it.
2. run 'net use * /delete' (this removes all mapped drives)
3. Back to cassowary Folder map and try again to map the path.(you may get an error but the drive actually maps with no problem)

**Comment by casualsnek** [comment]
> First of all thank you for this great project.
> 
> I figured out a way to fix it : TL;DR : remove all mapped drivers using powershell command and then remap. How to do it step by step:
> 
>     1. add powershell from guest apps in cassowary. and launch it.
> 
>     2. run 'net use * /delete' (this removes all mapped drives)
> 
>     3. Back to cassowary Folder map and try again to map the path.(you may get an error but the drive actually maps with no problem)

Thank you for the fix, i will try to find a good way to integrate this solution while creating shares !

**Comment by casualsnek** [comment]
> > First of all thank you for this great project.
> > I figured out a way to fix it : TL;DR : remove all mapped drivers using powershell command and then remap. How to do it step by step:
> > ```
> > 1. add powershell from guest apps in cassowary. and launch it.
> > 
> > 2. run 'net use * /delete' (this removes all mapped drives)
> > 
> > 3. Back to cassowary Folder map and try again to map the path.(you may get an error but the drive actually maps with no problem)
> > ```
> 
> Thank you for the fix, i will try to find a good way to integrate this solution while creating shares !

I was able to reproduce this error once, running ```net use * /delete``` fixed it and the error never showed up again ! Can i ask you if you could create shares properly after running the command once or had to run it again  ?

**Comment by CanasSimon** [comment]
After running the command for the first time I was able to map another drive the expected way (it even made Z: show up). The drives are for now still mapped correctly but I again cannot map new folders anymore, same error message.

**Comment by casualsnek** [comment]
> After running the command for the first time I was able to map another drive the expected way (it even made Z: show up). The drives are for now still mapped correctly but I again cannot map new folders anymore, same error message.

Thank you for reporting, i will add a workaround for this issue in next release !

**Comment by casualsnek** [comment]
I have pushed some changes for this. Can you test if this build can share directories properly: https://ufile.io/f/p5tod  ?
If this works and other few bugs are resolved i will make a new release

**Comment by casualsnek** [comment]
> I have pushed some changes for this. Can you test if this build can share directories properly: https://ufile.io/f/p5tod ? If this works and other few bugs are resolved i will make a new release

After installing the windows application you will need to restart vm for it to update !

**Comment by CanasSimon** [comment]
> I have pushed some changes for this. Can you test if this build can share directories properly: https://ufile.io/f/p5tod ? If this works and other few bugs are resolved i will make a new release
> 
> After installing the windows application you will need to restart vm for it to update !

It sadly doesn't seem to work :( And the workaround listed above doesn't work anymore either
The logs also say the exact same things as before

**Comment by casualsnek** [comment]
Is Z:\ drive still there ? If not can you manually run ```net use Z: \\tsclient\root``` on cmd window opened by ```python3 -m cassowary -c guest-run -- cmd.exe```.

This bug seems to affect randomly even when using same installation medium .

**Comment by casualsnek** [comment]
Looks like mapping works almost every time when any RDP application is active ! Opening explorer while creating a map fixes it ! Can anyone confirm if this works ? If it does i will start working on fix !

**Comment by casualsnek** [comment]
https://ufile.io/f/mtmpo

Can you test this build, installation instruction is same as previous build.
This build should fix this issue properly !

**Comment by CanasSimon** [comment]
It works! It gives me a Windows error message 'None is not a supported command', but it works!

**Comment by casualsnek** [comment]
> It works! It gives me a Windows error message 'None is not a supported command', but it works!

That was quick ! Thank you ! For None is supported error message you can open explorer navite to ```C:\Program Files\cassowary\`` open nowindow.vbs and edit line 6 so that it is ```command_line = ""``` instead. This should get rid of error message without rebooting. 
I will be pushing changes with nowindow.vbs fixed !

**Comment by casualsnek** [comment]
I just released v0.5. It should work without any manual workaround. If this works properly, i think i should close this issue !

**Comment by CanasSimon** [comment]
Everything is working as it should on my end

**Comment by Le0xFF** [comment]
Here's the linux log!
[casualrdh.log](https://github.com/casualsnek/cassowary/files/8042967/casualrdh.log)

**Comment by casualsnek** [comment]
> Yes, always the same error 87!
> 
> Here's the linux log! [casualrdh.log](https://github.com/casualsnek/cassowary/files/8042967/casualrdh.log)

Was any windows application running through cassowary when you tried to create share ?

**Comment by Le0xFF** [comment]
Little update, it seems that even the previous error, the folder map is created, tho is not visible on the main cassowary gui.
(`P:` is the drive I created)
![NotError](https://user-images.githubusercontent.com/43521247/153465283-b5203f9e-154b-416e-94fc-532e8452d0b3.png)

**Comment by casualsnek** [comment]
Can you switch tabs on folder mapping and see if it appears there ?  
From your log file it looked like you had some windows app running while you created share. Am i correct  ? If not i can see where error happened !

**Comment by Le0xFF** [comment]
Switching tabs does not make appear any of my mapped device, but they are created.
I can see them in explorer, even if cassowary gui thrown an error.

**Comment by Le0xFF** [comment]
Here's the config from linux!
[config.json.log](https://github.com/casualsnek/cassowary/files/8043344/config.json.log)

**Comment by casualsnek** [comment]
Oops log says:  Unknown error while trying to map to drive 'Esecuzione comando riuscita. . seems to mean Command execution successful. 
And it was returned as error message, i need to make sure it is language agnostic 😆

**Comment by Le0xFF** [comment]
It seems like I can't install the `.whl` file on linux?
It says `ERROR: cassowary-0.5-py3-tmpbuild.whl is not a valid wheel filename.`

I installed the `.zip` on Windows and no error at all.
Just one minor error when it says `Setting up path variable`, it throw up an error about `Invalid syntax` of `SETX` command, but it already happened with previous release, so I don't know if that's a major problem or not.
![SyntaxError](https://user-images.githubusercontent.com/43521247/153706923-ea402ca8-12f4-40bb-b969-eb36d3f2b847.png)

**Comment by casualsnek** [comment]
> It seems like I can't install the `.whl` file on linux?
> It says `ERROR: cassowary-0.5-py3-tmpbuild.whl is not a valid wheel filename.`
> 
> I installed the `.zip` on Windows and no error at all.
> Just one minor error when it says `Setting up path variable`, it throw up an error about `Invalid syntax` of `SETX` command, but it already happened with previous release, so I don't know if that's a major problem or not.
> ![SyntaxError](https://user-images.githubusercontent.com/43521247/153706923-ea402ca8-12f4-40bb-b969-eb36d3f2b847.png)
>  

The error message on windows is expected during upgrade, for the linux whl file i will update this comment with another one !

**Comment by Le0xFF** [comment]
Okay, both installed, but same behaviour: I get an error, the map is created but I don't see the new mapped drive in the GUI.
![NewError](https://user-images.githubusercontent.com/43521247/153713522-64d8675c-7ce3-4a56-9f0d-d68340b43a7b.png)

[linux_casualrdh.log](https://github.com/casualsnek/cassowary/files/8053480/linux_casualrdh.log)
[windows_casualrdh.log](https://github.com/casualsnek/cassowary/files/8053481/windows_casualrdh.log)

**Comment by casualsnek** [comment]
> Okay, both installed, but same behaviour: I get an error, the map is created but I don't see the new mapped drive in the GUI. ![NewError](https://user-images.githubusercontent.com/43521247/153713522-64d8675c-7ce3-4a56-9f0d-d68340b43a7b.png)
> 
> [linux_casualrdh.log](https://github.com/casualsnek/cassowary/files/8053480/linux_casualrdh.log) [windows_casualrdh.log](https://github.com/casualsnek/cassowary/files/8053481/windows_casualrdh.log)

https://ufile.io/m8kapmig This build should work !

**Comment by Le0xFF** [comment]
Okay I installed it, I removed the network shares and now I get a new error!
This time the `Z:` drive is not mounted automatically tho.
![NewError](https://user-images.githubusercontent.com/43521247/153749406-c3f8f5e0-b596-437c-b597-9cd8791c68b0.png)

**Comment by Le0xFF** [comment]
Yay, it's finally working!
I guess `Z:` is supposed to not show up in the GUI as a mapped drive, right?

![Working](https://user-images.githubusercontent.com/43521247/153767467-066bf8e5-fd8c-4b60-bec1-bac55677f68c.png)

Just in case, here are the logs!
[linux_casualrdh.log](https://github.com/casualsnek/cassowary/files/8056251/linux_casualrdh.log)
[windows_casualrdh.log](https://github.com/casualsnek/cassowary/files/8056252/windows_casualrdh.log)

**Comment by casualsnek** [comment]
> Yay, it's finally working! I guess `Z:` is supposed to not show up in the GUI as a mapped drive, right?
> 
> ![Working](https://user-images.githubusercontent.com/43521247/153767467-066bf8e5-fd8c-4b60-bec1-bac55677f68c.png)
> 
> Just in case, here are the logs! [linux_casualrdh.log](https://github.com/casualsnek/cassowary/files/8056251/linux_casualrdh.log) [windows_casualrdh.log](https://github.com/casualsnek/cassowary/files/8056252/windows_casualrdh.log)

Yes, Z is intended to be hidden from gui and gets recreated everytime RDP session starts !

**Comment by casualsnek** [comment]
Thank you for helping with all the testing, if you don't encounter this issue for next 1 week i will close this issue thread !

**Comment by TheWranger** [comment]
> Okay I installed it, I removed the network shares and now I get a new error! This time the `Z:` drive is not mounted automatically tho. I still can't see any mapped device in the GUI. ![NewError](https://user-images.githubusercontent.com/43521247/153749406-c3f8f5e0-b596-437c-b597-9cd8791c68b0.png)
> 
> [linux_casualrdh.log](https://github.com/casualsnek/cassowary/files/8055358/linux_casualrdh.log)
> 
> Oh btw the `Windows -> Linux` mapping tab is working now! But when I mount the folder, the **Status** column is not updated, it always says `Not Mounted`.

Hey, how did you fix this missing parameter error? Thanks

**Comment by ghost** [comment]
> > Okay I installed it, I removed the network shares and now I get a new error! This time the `Z:` drive is not mounted automatically tho. I still can't see any mapped device in the GUI. ![NewError](https://user-images.githubusercontent.com/43521247/153749406-c3f8f5e0-b596-437c-b597-9cd8791c68b0.png)
> > 
> > [linux_casualrdh.log](https://github.com/casualsnek/cassowary/files/8055358/linux_casualrdh.log)
> > 
> > Oh btw the `Windows -> Linux` mapping tab is working now! But when I mount the folder, the **Status** column is not updated, it always says `Not Mounted`.
> 
> Hey, how did you fix this missing parameter error? Thanks

To be honest, it's been a while since the last time I used it, so I really don't remember anymore.
Try with latest build? Or you can always join dev's discord and ask him directly!
